# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
      * The paper explores the concept of data tidying, which is when data is restructured to make analysis easier. 
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
      * The tidy data standard has two goals: simplify data cleaning and to facilitate data analysis and tool development. Having a standard makes it easier to clean data because people don't have to "reinvent the wheel" each time and can save brainpower.
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
      * The first sentence about families is referring to the Leo Tolstoy quote, and means that tidy datasets adhere to a standardized structure and format, which makes them similar and consistent in their organization. In contrast, messy datasets vary widely in terms of how they are structured and formatted. There are many ways to be a messy dataset (or an unhappy family) but not many ways to be tidy (or happy).
      * The second sentence means that when you look at a specific dataset, it's often straightforward to identify what makes up individual data points (observations) and what represents the characteristics being measured (variables). However, when attempting to create a general, abstract definition that applies to all datasets, it becomes challenging because the concept of what makes a variable and an observation can change based on the context and dataset.
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
      * Values, according to Wickham, are the individual data points in a dataset, and they can be either quantitative (numbers) or qualitative (strings). They represent measurements or attributes of interest.
      * Variables are attributes or characteristics being measured or observed in a dataset. Variables group together all values that measure the same underlying attribute across different units or observations. 
      * Observations represent the individual units or entities being studied in a dataset. Each observation contains values measured on the same unit across various attributes or variables.
  5. How is "Tidy Data" defined in section 2.3?
      * Tidy data is defined as a standardized way of structuring a dataset, aligning data's meaning with its structure:
        - Each variable becomes a column.
        - Each observation becomes a row.
        - Each observational unit forms a separate table.
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
      * The 5 most common problems are:
        - Column headers as values, not variable names
        - Multiple variables stored in one column
        - Variables stored in both rows and columns
        - Multiple types of observational units stored in the same table
        - A single observational unit is stored in multiple tables
      * Table 4 is messy because it demonstrates the first problem, where column headers contain values instead of variable names. 
      * "Melting" a dataset refers to the process of transforming it from a wide format, where variables are represented as columns, into a long format, where values are stacked into rows. 
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
        * Table 11 is considered messy because it lacks the structured organization required for efficient data analysis, while Table 12 is tidy and "molten" because it has been transformed into a format that conforms to the principles of tidy data, making it more suitable for analysis and interpretation.
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?
       * The "chicken-and-egg" issue in data wrangling is where the structure of data (tidy data) and the tools used for analysis (tidy tools) are interdependent. Tidy data is most beneficial when used with compatible tools, creating a challenge of mutual reliance. This situation can limit progress, as advancements in data analysis may be constrained by existing data structures and tools.
      * Wickham anticipates future improvements in data wrangling, both incrementally and through exploring alternative data structuring methods. He emphasizes the importance of considering human factors and user-centered design in tool development and envisions the tidy data framework evolving as more research is conducted in this area.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

**Q2.1**
To clean the Price column, I removed commas and then converted it to numeric. I end up with 0 missing values.

In [2]:
import pandas as pd

# load dataset
airbnb_df = pd.read_csv('./data/airbnb_hw.csv')

# remove commas from col
airbnb_df['Price'] = airbnb_df['Price'].str.replace(',', '')

# convert col to numeric type
airbnb_df['Price'] = pd.to_numeric(airbnb_df['Price'], errors='coerce')

# count num of missing values in price col
missing_val_count = airbnb_df['Price'].isna().sum()

missing_val_count


0

**Q2.2**
To clean the Type variable, I combined all of the boating related values to be labeled as "Boating". I also combined all of the values that I felt were variations of "Unknown" and combined them as such.

In [3]:
# load dataset
sharks_df = pd.read_csv('./data/sharks.csv')

# display unique values
unique_types = sharks_df['Type'].unique()
unique_types

C:\Users\allard.ti\AppData\Local\Temp\ipykernel_36856\610896713.py:2: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sharks_df = pd.read_csv('./data/sharks.csv')


array(['Unprovoked', 'Provoked', 'Questionable', 'Watercraft',
       'Unconfirmed', 'Unverified', 'Invalid', 'Under investigation',
       'Boating', 'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [4]:
# combine 'boating' 'watercraft' 'boat' 'boatomg' 'sea disaster'
sharks_df['Type'] = sharks_df['Type'].replace(['Watercraft', 'Boat', 'Boatomg', 'Sea Disaster'], 'Boating')

# combine 'unconfirmed' 'unverified' 'questionable' 'under investigation' 'invalid' to 'unknown'
sharks_df['Type'] = sharks_df['Type'].replace(['Unconfirmed', 'Unverified', 'Questionable', 'Under investigation', 'Invalid'], 'Unknown')

# replace missing vals with unknown
sharks_df['Type'].fillna('Unknown', inplace=True)

# display updated unique values
updated_unique_types = sharks_df['Type'].unique()
updated_unique_types

array(['Unprovoked', 'Provoked', 'Unknown', 'Boating'], dtype=object)

**Q2.3**


In [7]:
import numpy as np

# load dataset
pretrial_df = pd.read_csv('./data/pretrial_data.csv')

pretrial_df['released'].head()

0    NaN
1    0.0
2    0.0
3    0.0
4    1.0
Name: released, dtype: float64

In [8]:
# Check the unique values and their counts in the 'released' column
unique_vals = pretrial_df['released'].value_counts(dropna=False)
unique_vals


1.0    19154
0.0     3801
NaN       31
Name: released, dtype: int64

**Q2.4**
For rows where `sentence_type` was 0.0 (meaning probation/no incarceration) but did not match `sentence`, I set `sentence` to 0.

In [9]:
pretrial_df.head()

Unnamed: 0 case_type   age sex race  is_poor  bond  bond_type  prior_F  \
0           0         F  31.0   M    W      NaN   NaN          7      0.0   
1           1         F  60.0   M    B      NaN   NaN          7     13.0   
2           2         M  27.0   M    W      NaN   NaN          7      0.0   
3           3         M  27.0   M    B      0.0   NaN          7      0.0   
4           4         F  28.0   F    W      0.0   NaN          7      0.0   

   prior_M  gini  released  sentence_type   sentence  bond_NA  held_wo_bail  \
0      0.0  0.44       NaN            NaN        NaN     True         False   
1     21.0  0.44       0.0            0.0  60.000000     True         False   
2      0.0  0.44       0.0            1.0  12.000000     True         False   
3      9.0  0.44       0.0            1.0   0.985626     True         False   
4      0.0  0.44       1.0            4.0   0.000000     True         False   

   sentence_NA  
0         True  
1        False  
2        False  
3        False  
4         True

In [10]:
# set the `sentence` value to 0 for records with `sentence_type` as "Probation/no incarceration"
pretrial_df.loc[pretrial_df['sentence_type'] == 0, 'sentence'] = 0

In [11]:
pretrial_df.head()

Unnamed: 0 case_type   age sex race  is_poor  bond  bond_type  prior_F  \
0           0         F  31.0   M    W      NaN   NaN          7      0.0   
1           1         F  60.0   M    B      NaN   NaN          7     13.0   
2           2         M  27.0   M    W      NaN   NaN          7      0.0   
3           3         M  27.0   M    B      0.0   NaN          7      0.0   
4           4         F  28.0   F    W      0.0   NaN          7      0.0   

   prior_M  gini  released  sentence_type   sentence  bond_NA  held_wo_bail  \
0      0.0  0.44       NaN            NaN        NaN     True         False   
1     21.0  0.44       0.0            0.0   0.000000     True         False   
2      0.0  0.44       0.0            1.0  12.000000     True         False   
3      9.0  0.44       0.0            1.0   0.985626     True         False   
4      0.0  0.44       1.0            4.0   0.000000     True         False   

   sentence_NA  
0         True  
1        False  
2        False  
3        False  
4         True

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

In [21]:
# 1 load data
college_df = pd.read_csv('./data/college_completion.csv')


In [19]:
# 2 dimensions
dim = college_df.shape
num_obsv = dim[0]
college_vars = college_df.columns

dim, num_obsv, college_vars, college_df.head()


((3798, 63),
 3798,
 Index(['index', 'unitid', 'chronname', 'city', 'state', 'level', 'control',
        'basic', 'hbcu', 'flagship', 'long_x', 'lat_y', 'site', 'student_count',
        'awards_per_value', 'awards_per_state_value', 'awards_per_natl_value',
        'exp_award_value', 'exp_award_state_value', 'exp_award_natl_value',
        'exp_award_percentile', 'ft_pct', 'fte_value', 'fte_percentile',
        'med_sat_value', 'med_sat_percentile', 'aid_value', 'aid_percentile',
        'endow_value', 'endow_percentile', 'grad_100_value',
        'grad_100_percentile', 'grad_150_value', 'grad_150_percentile',
        'pell_value', 'pell_percentile', 'retain_value', 'retain_percentile',
        'ft_fac_value', 'ft_fac_percentile', 'vsa_year',
        'vsa_grad_after4_first', 'vsa_grad_elsewhere_after4_first',
        'vsa_enroll_after4_first', 'vsa_enroll_elsewhere_after4_first',
        'vsa_grad_after6_first', 'vsa_grad_elsewhere_after6_first',
        'vsa_enroll_after6_first', 'vsa_

In [22]:
# 3 cross tabulate control and level
cross_tab = pd.crosstab(college_df['control'], college_df['level'])

cross_tab

level                   2-year  4-year
control                               
Private for-profit         465     527
Private not-for-profit      68    1180
Public                     926     632

**Q3.3**
Here I noticed that Private not-for-profit colleges are mostly 4-year colleges. Additionally, there are more Public 2-years than 4-years. Lastly, Private for-profit and not-for-profit are about evenly distributed.

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.